// agentfmt/dist/index.js
import pc from 'picocolors';
// === Status indicators ===
export const ok = (msg) => `${pc.green('✓')} ${msg}`;
export const fail = (msg) => `${pc.red('✗')} ${msg}`;
export const warn = (msg) => `${pc.yellow('⚠')} ${msg}`;
export const info = (msg) => `${pc.cyan('ℹ')} ${msg}`;
// === Colors (optional, re-export for convenience) ===
export const dim = pc.dim;
export const bold = pc.bold;
export const green = pc.green;
export const red = pc.red;
export const yellow = pc.yellow;
export const cyan = pc.cyan;
export const gray = pc.gray;
export const magenta = pc.magenta;
export const blue = pc.blue;
export const white = pc.white;
// === Primitives ===
export function kv(key, value) {
    if (value === null || value === undefined)
        return '';
    return `${key}: ${value}`;
}
export function box(width, height, x, y) {
    const size = `${Math.round(width)}×${Math.round(height)}`;
    if (x !== undefined && y !== undefined) {
        return `${size} at (${Math.round(x)}, ${Math.round(y)})`;
    }
    return size;
}
export function entity(type, name, id) {
    const idPart = id ? ` (${id})` : '';
    return `[${type}] "${name}"${idPart}`;
}
export function truncate(text, maxLen = 40, ellipsis = '…') {
    if (text.length <= maxLen)
        return text;
    return text.slice(0, maxLen - ellipsis.length) + ellipsis;
}
export function quoted(text, maxLen) {
    const t = maxLen ? truncate(text, maxLen) : text;
    return `"${t.replace(/\n/g, '↵')}"`;
}
function formatDetails(details, indent) {
    if (Array.isArray(details)) {
        return details.filter(Boolean).map((d) => `${indent}${d}`);
    }
    return Object.entries(details)
        .filter(([, v]) => v !== null && v !== undefined && v !== '')
        .map(([k, v]) => `${indent}${k}: ${v}`);
}
export function list(items, options = {}) {
    const { numbered = true, start = 0, compact = false } = options;
    const lines = [];
    const maxIndex = start + items.length - 1;
    const indexWidth = numbered ? `[${maxIndex}]`.length : 2;
    for (let i = 0; i < items.length; i++) {
        const item = items[i];
        const prefix = numbered ? `[${start + i}]`.padEnd(indexWidth) : '•';
        const indent = ' '.repeat(indexWidth + 1);
        lines.push(`${prefix} ${item.header}`);
        if (item.details) {
            const detailLines = formatDetails(item.details, indent);
            lines.push(...detailLines);
        }
        // Add blank line between items unless compact mode or no details
        if (!compact && item.details && i < items.length - 1) {
            lines.push('');
        }
    }
    return lines.join('\n');
}
function formatTreeNode(node, depth, index, options) {
    const { maxDepth = -1, showIndex = true } = options;
    if (maxDepth !== -1 && depth > maxDepth)
        return [];
    const lines = [];
    const indent = '  '.repeat(depth);
    const indexPart = showIndex ? `[${index}] ` : '';
    lines.push(`${indent}${indexPart}${node.header}`);
    if (node.details) {
        const detailIndent = indent + '    ';
        const detailLines = formatDetails(node.details, '');
        if (detailLines.length > 0) {
            lines.push(`${detailIndent}${detailLines.join(' | ')}`);
        }
    }
    if (node.children && (maxDepth === -1 || depth < maxDepth)) {
        for (let i = 0; i < node.children.length; i++) {
            lines.push(...formatTreeNode(node.children[i], depth + 1, i, options));
        }
    }
    return lines;
}
export function tree(root, options = {}) {
    return formatTreeNode(root, 0, 0, options).join('\n');
}
// === Props block ===
export function props(data, indent = '  ') {
    return Object.entries(data)
        .filter(([, v]) => v !== null && v !== undefined && v !== '')
        .map(([k, v]) => `${indent}${k}: ${v}`)
        .join('\n');
}
export function histogram(items, options = {}) {
    const { maxBarLength = 30, scale = 10, unit = '×' } = options;
    return items
        .map((item) => {
        const bar = '█'.repeat(Math.min(Math.ceil(item.value / scale), maxBarLength));
        const tagPart = item.tag ? ` (${item.tag})` : '';
        const suffixPart = item.suffix ? ` ${item.suffix}` : '';
        return `${item.label}  ${bar} ${item.value}${unit}${tagPart}${suffixPart}`;
    })
        .join('\n');
}
// === Summary line ===
export function summary(counts) {
    return Object.entries(counts)
        .filter(([, v]) => v > 0)
        .map(([k, v]) => `${v} ${k}`)
        .join(', ');
}
export function node(data, extraDetails) {
    const lines = [];
    lines.push(entity(data.type.toLowerCase(), data.name, data.id));
    const details = [];
    if (data.width !== undefined && data.height !== undefined) {
        details.push(`box: ${box(data.width, data.height, data.x, data.y)}`);
    }
    if (extraDetails) {
        for (const [k, v] of Object.entries(extraDetails)) {
            if (v !== null && v !== undefined && v !== '') {
                details.push(`${k}: ${v}`);
            }
        }
    }
    for (const detail of details) {
        lines.push(`  ${detail}`);
    }
    return lines.join('\n');
}
const SEVERITY_ICONS = {
    error: '✖',
    warning: '⚠',
    info: 'ℹ'
};
export function lint(groups, options = {}) {
    const { verbose = false } = options;
    const lines = [];
    for (const group of groups) {
        const hasError = group.messages.some((m) => m.severity === 'error');
        const icon = hasError ? pc.red('✖') : pc.yellow('⚠');
        lines.push(`${icon} ${pc.bold(group.path)}`);
        for (const msg of group.messages) {
            const sevIcon = msg.severity === 'error'
                ? pc.red(SEVERITY_ICONS.error)
                : msg.severity === 'warning'
                    ? pc.yellow(SEVERITY_ICONS.warning)
                    : pc.cyan(SEVERITY_ICONS.info);
            const rulePart = msg.rule ? `  ${pc.dim(msg.rule)}` : '';
            lines.push(`    ${sevIcon}  ${msg.message}${rulePart}`);
            if (verbose && msg.suggest) {
                lines.push(`       ${pc.dim(`→ ${msg.suggest}`)}`);
            }
        }
        lines.push('');
    }
    return lines.join('\n');
}
export function lintSummary(counts) {
    const parts = [];
    if (counts.errors > 0) {
        parts.push(pc.red(`${counts.errors} error${counts.errors !== 1 ? 's' : ''}`));
    }
    if (counts.warnings > 0) {
        parts.push(pc.yellow(`${counts.warnings} warning${counts.warnings !== 1 ? 's' : ''}`));
    }
    if (counts.infos && counts.infos > 0) {
        parts.push(pc.cyan(`${counts.infos} info`));
    }
    if (parts.length === 0) {
        return pc.green('✔ No issues found');
    }
    return '─'.repeat(50) + '\n' + parts.join('  ');
}
