// figma-use/dist/cli/render-X2QODZOQ.js
import{a as f,b as u,c as x,d as g,e as y}from"./chunk-A2CPD4PM.js";import{b as m,c as d}from"./chunk-5VFGFCBY.js";import{g as c,i as a}from"./chunk-3VLTU5MO.js";import{a as l,b as r,n as h}from"./chunk-77YLI2U6.js";import"./chunk-7ZS3Q2QA.js";import"./chunk-5J6HXUF6.js";import{defineCommand as w}from"citty";import*as S from"esbuild";import{existsSync as R}from"fs";import{resolve as b}from"path";h();async function E(){let e=[];for await(let t of process.stdin)e.push(t);return Buffer.concat(e).toString("utf-8")}var F=`
  export const Frame = 'frame'
  export const Text = 'text'
  export const Rectangle = 'rectangle'
  export const Ellipse = 'ellipse'
  export const Line = 'line'
  export const Image = 'image'
  export const SVG = 'svg'
  export const Icon = (props) => ({ __icon: true, name: props.name, size: props.size, color: props.color })
  export const View = 'frame'
  export const Rect = 'rectangle'
  export const Section = 'section'
  export const Group = 'group'
  export const defineComponent = (name, el) => () => el
  export const defineVars = (vars) => Object.fromEntries(Object.entries(vars).map(([k, v]) => [k, v.value]))
`;async function N(e){let t=e.trim();!t.includes("import ")&&!t.includes("export ")?t=`import { Frame, Text, Rectangle, Ellipse, Line, Image, SVG, Icon } from 'figma-use/render'
export default () => ${t}`:t.includes("export ")||(t=`${t}
export default () => null`);let i=(await S.build({stdin:{contents:t,loader:"tsx"},bundle:!0,write:!1,format:"iife",globalName:"__mod",jsx:"transform",jsxFactory:"React.createElement",plugins:[{name:"mock-imports",setup(n){n.onResolve({filter:/^figma-use\/render$|^\./},o=>({path:o.path,namespace:"mock"})),n.onLoad({filter:/.*/,namespace:"mock"},()=>({contents:F,loader:"js"}))}}]})).outputFiles[0].text;return new Function("React",`${i}; return __mod.default`)(u)}var v=`
Render JSX to Figma.

EXAMPLES

  echo '<frame style={{ p: 24, bg: "#3B82F6", rounded: 12 }}>
    <text style={{ size: 18, color: "#FFF" }}>Hello</text>
  </frame>' | figma-use render --stdin

  figma-use render ./Card.figma.tsx --props '{"title": "Hello"}'

ELEMENTS

  Frame, Rectangle, Ellipse, Text, Line, Star, Polygon, Vector, Group, Icon, Image

SHORTHANDS

  w, h          width, height
  bg            fill color
  rounded       cornerRadius
  p, px, py     padding
  flex          direction ("row" | "col")
  gap           spacing
  wrap          enable flex wrap
  size, weight  fontSize, fontWeight
  justify, items  alignment
`,_=w({meta:{description:"Render JSX to Figma"},args:{examples:{type:"boolean",description:"Show examples"},file:{type:"positional",description:"TSX/JSX file",required:!1},stdin:{type:"boolean",description:"Read from stdin"},props:{type:"string",description:"JSON props"},parent:{type:"string",description:"Parent node ID"},x:{type:"string",description:"X position"},y:{type:"string",description:"Y position"},export:{type:"string",description:"Named export"},json:{type:"boolean",description:"JSON output"}},async run({args:e}){if(e.examples){console.log(v);return}try{let t;if(e.stdin){let o=await E();o.trim()||(console.error(r("No input from stdin")),process.exit(1)),t=await N(o)}else if(e.file){let o=b(e.file);R(o)||(console.error(r(`File not found: ${o}`)),process.exit(1)),t=(await import(o))[e.export||"default"],t||(console.error(r(`Export "${e.export||"default"}" not found`)),process.exit(1))}else console.error(r("Provide file or --stdin")),process.exit(1);if(!y())try{let o=await c("get-variables",{simple:!0});g(o)}catch{}let p=e.props?JSON.parse(e.props):{},i=f(t,p),s=d(i);s.length>0&&(e.json||console.log(`Preloading ${s.length} icon(s)...`),await m(s));let n=await x(i,{parent:e.parent,x:e.x?Number(e.x):void 0,y:e.y?Number(e.y):void 0});e.json?console.log(JSON.stringify(n,null,2)):(console.log(l(`Rendered: ${n.id}`)),console.log(`  name: ${n.name}`))}catch(t){a(t)}}});export{_ as default};
